File: filament/src/Fence.cpp
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "details/Fence.h"

#include "details/Engine.h"

#include <filament/Fence.h>

#include <utils/Panic.h>

namespace filament {

using namespace backend;

namespace details {


utils::Mutex FFence::sLock;
utils::Condition FFence::sCondition;

static const constexpr uint64_t PUMP_INTERVAL_MILLISECONDS = 1;

using ms = std::chrono::milliseconds;
using ns = std::chrono::nanoseconds;

FFence::FFence(FEngine& engine, Type type)
    : mEngine(engine), mFenceSignal(std::make_shared<FenceSignal>(type)) {
    DriverApi& driverApi = engine.getDriverApi();
    if (type == Type::HARD) {
        mFenceHandle = driverApi.createFence();
    }

    // we have to first wait for the fence to be signaled by the command stream
    auto& fs = mFenceSignal;
    driverApi.queueCommand([fs]() {
        fs->signal();
    });
}

void FFence::terminate(FEngine& engine) noexcept {
    FenceSignal * const fs = mFenceSignal.get();
    fs->signal(FenceSignal::DESTROYED);
    if (fs->mType == Type::HARD) {
        if (mFenceHandle) {
            FEngine::DriverApi& driver = engine.getDriverApi();
            driver.destroyFence(mFenceHandle);
            mFenceHandle.clear();
        }
    }
}

UTILS_NOINLINE
FenceStatus FFence::waitAndDestroy(FFence* fence, Mode mode) noexcept {
    assert(fence);
    FenceStatus status = fence->wait(mode, FENCE_WAIT_FOR_EVER);
    fence->mEngine.destroy(fence);
    return status;
}

UTILS_NOINLINE
FenceStatus FFence::wait(Mode mode, uint64_t timeout) noexcept {
    ASSERT_PRECONDITION(UTILS_HAS_THREADING || timeout == 0, "Non-zero timeout requires threads.");

    FEngine& engine = mEngine;

    if (mode == Mode::FLUSH) {
        engine.flush();
    }

    FenceSignal * const fs = mFenceSignal.get();

    FenceStatus status;

    if (UTILS_LIKELY(!engine.pumpPlatformEvents())) {
        status = fs->wait(timeout);
    } else {
        // Unfortunately, some platforms might force us to have sync points between the GL thread
        // and user thread. To prevent deadlock on these platforms, we chop up the waiting time into
        // polling and pumping the platform's event queue.
        const auto startTime = std::chrono::system_clock::now();
        while (true) {
            status = fs->wait(ns(ms(PUMP_INTERVAL_MILLISECONDS)).count());
            if (status != FenceStatus::TIMEOUT_EXPIRED) {
                break;
            }
            engine.pumpPlatformEvents();
            const auto elapsed = std::chrono::system_clock::now() - startTime;
            if (timeout != Fence::FENCE_WAIT_FOR_EVER && elapsed >= ns(timeout)) {
                break;
            }
        }
    }

    if (status != FenceStatus::CONDITION_SATISFIED) {
        return status;
    }

    if (fs->mType == Type::HARD) {
        // note: even if the driver doesn't support h/w fences, mFenceHandle will be valid
        status = engine.getDriverApi().wait(mFenceHandle, timeout);
    }
    return status;
}

UTILS_NOINLINE
void FFence::FenceSignal::signal(State s) noexcept {
    std::unique_lock<utils::Mutex> lock(FFence::sLock);
    mState = s;
    lock.unlock();
    FFence::sCondition.notify_all();
}

UTILS_NOINLINE
Fence::FenceStatus FFence::FenceSignal::wait(uint64_t timeout) noexcept {
    std::unique_lock<utils::Mutex> lock(FFence::sLock);
    while (mState == UNSIGNALED) {
        if (mState == DESTROYED) {
            return FenceStatus::ERROR;
        }
        if (timeout == FENCE_WAIT_FOR_EVER) {
            FFence::sCondition.wait(lock);
        } else {
            if (timeout == 0 ||
                    sCondition.wait_for(lock, ns(timeout)) == std::cv_status::timeout) {
                return FenceStatus::TIMEOUT_EXPIRED;
            }
        }
    }
    return FenceStatus::CONDITION_SATISFIED;
}

// ------------------------------------------------------------------------------------------------
} // namespace details

// ------------------------------------------------------------------------------------------------
// Trampoline calling into private implementation
// ------------------------------------------------------------------------------------------------

using namespace details;

FenceStatus Fence::waitAndDestroy(Fence* fence, Mode mode) {
    return FFence::waitAndDestroy(upcast(fence), mode);
}

FenceStatus Fence::wait(Mode mode, uint64_t timeout) {
    return upcast(this)->wait(mode, timeout);
}


} // namespace filament

File: android/filament-android/src/main/java/com/google/android/filament/Fence.java
package com.google.android.filament;
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import androidx.annotation.NonNull;

public class Fence {
    private long mNativeObject;

    Fence(long nativeFence) {
        mNativeObject = nativeFence;
    }

    public static final long WAIT_FOR_EVER = -1;

    public enum Mode {
        FLUSH,
        DONT_FLUSH
    }

    public enum FenceStatus {
        ERROR,
        CONDITION_SATISFIED,
        TIMEOUT_EXPIRED
    }

    /**
     * Client-side wait on the Fence.
     *
     * Blocks the current thread until the Fence signals.
     *
     * @param mode      Whether the command stream is flushed before waiting or not.
     * @param timeoutNanoSeconds   Wait time out in nanoseconds. Using a timeout of 0 is a way to query the state of the fence.
     *                  A timeout value of WAIT_FOR_EVER is used to disable the timeout.
     * @return          FenceStatus::CONDITION_SATISFIED on success,
     *                  FenceStatus::TIMEOUT_EXPIRED if the time out expired or
     *                  FenceStatus::ERROR in other cases.
     * @throws Error if the backend thread encountered an unrecoverable error.
     */
    public FenceStatus wait(@NonNull Mode mode, long timeoutNanoSeconds) {
        int nativeResult = nWait(getNativeObject(), mode.ordinal(), timeoutNanoSeconds);
        switch (nativeResult) {
            case -1:
                return FenceStatus.ERROR;
            case 0:
                return FenceStatus.CONDITION_SATISFIED;
            case 1:
                return FenceStatus.TIMEOUT_EXPIRED;
            default:
                // this should never happend
                return FenceStatus.ERROR;
        }
    }

    /**
     * Client-side wait on a Fence and destroy the Fence.
     *
     * @param fence Fence object to wait on.
     * @param mode  Whether the command stream is flushed before waiting or not.
     * @return  FenceStatus::CONDITION_SATISFIED on success,
     *          FenceStatus::ERROR otherwise.
     * @throws Error if the backend thread encountered an unrecoverable error.
     */
    public static FenceStatus waitAndDestroy(@NonNull Fence fence, @NonNull Mode mode) {
        int nativeResult = nWaitAndDestroy(fence.getNativeObject(), mode.ordinal());
        switch (nativeResult) {
            case -1:
                return FenceStatus.ERROR;
            case 0:
                return FenceStatus.CONDITION_SATISFIED;
            default:
                // this should never happen
                return FenceStatus.ERROR;
        }
    }

    public long getNativeObject() {
        if (mNativeObject == 0) {
            throw new IllegalStateException("Calling method on destroyed Fence");
        }
        return mNativeObject;
    }

    void clearNativeObject() {
        mNativeObject = 0;
    }

    private static native int nWait(long nativeFence, int mode, long timeoutNanoSeconds);
    private static native int nWaitAndDestroy(long nativeFence, int mode);
}
